UpdateService/handler.ts

```diff
@@ -0,0 +1,189 @@
+
+import { Context } from "@azure/functions";
+import { identity } from "fp-ts/lib/function";
+import { TaskEither } from "fp-ts/lib/TaskEither";
+import { ServiceModel } from "io-functions-commons/dist/src/models/service";
+import { ContextMiddleware } from "io-functions-commons/dist/src/utils/middlewares/context_middleware";
+import { RequiredParamMiddleware } from "io-functions-commons/dist/src/utils/middlewares/required_param";
+import { NonEmptyString } from "italia-ts-commons/lib/strings";
+import { APIClient } from "../clients/admin";
+import { Service } from "../generated/api-admin/Service";
+import { SubscriptionKeys } from "../generated/api-admin/SubscriptionKeys";
+import { ServicePayload } from "../generated/definitions/ServicePayload";
+import { ServiceWithSubscriptionKeys } from "../generated/definitions/ServiceWithSubscriptionKeys";
+import { withApiRequestWrapper } from "../utils/api";
+import { getLogger, ILogger } from "../utils/logging";
+import { ErrorResponses, IResponseErrorUnauthorized } from "../utils/responses";
+import { serviceOwnerCheckTask } from "../utils/subscription";
+
+type ResponseTypes =
+  | IResponseSuccessJson<ServiceWithSubscriptionKeys>
+  | IResponseErrorUnauthorized
+  | IResponseErrorForbiddenNotAuthorized
+  | IResponseErrorNotFound
+  | IResponseErrorTooManyRequests
+  | IResponseErrorInternal;
+
+/**
+ * Type of a UpdateService handler.
+ *
+ * UpdateService expects a service_id and a service payload as input
+ * and returns updated service with subscription keys
+ */
+type IUpdateServiceHandler = (
+  context: Context,
+  auth: IAzureApiAuthorization,
+  clientIp: ClientIp,
+  attrs: IAzureUserAttributes,
+  serviceId: NonEmptyString,
+  servicePayload: ServicePayload
+) => Promise<ResponseTypes>;
+
+const logPrefix = "UpdateServiceHandler";
+
+const getSubscriptionKeysTask = (
+  logger: ILogger,
+  apiClient: APIClient,
+  serviceId: NonEmptyString
+): TaskEither<ErrorResponses, SubscriptionKeys> =>
+  withApiRequestWrapper(
+    logger,
+    () =>
+      apiClient.getSubscriptionKeys({
+        service_id: serviceId
+      }),
+    200
+  );
+
+const getServiceTask = (
+  logger: ILogger,
+  apiClient: APIClient,
+  serviceId: NonEmptyString
+): TaskEither<ErrorResponses, Service> =>
+  withApiRequestWrapper(
+    logger,
+    () =>
+      apiClient.getService({
+        service_id: serviceId
+      }),
+    200
+  );
+
+const updateServiceTask = (
+  logger: ILogger,
+  apiClient: APIClient,
+  servicePayload: ServicePayload,
+  serviceId: NonEmptyString,
+  retrievedService: Service
+): TaskEither<ErrorResponses, Service> =>
+  withApiRequestWrapper(
+    logger,
+    () =>
+      apiClient.updateService({
+        body: {
+          ...retrievedService,
+          ...servicePayload,
+          service_id: serviceId // TODO insert check on ADB2C token_name from Active Directory see: https://www.pivotaltracker.com/story/show/174823724
```

Comment:
Technically this is not required because it is already included in "...retrivedService".

Reply:
yes but in this way we could update existing services _after_ the token name is assigned to the user account (otherwise it will remain null for previously created services)